dataprofiler/data_readers/csv_data.py

```diff
@@ -104,21 +104,114 @@ def _determine_delimiter_of_str(data_as_str):
+        :type str_thresh: float
+        :return: index for row estimated to be the last valid header
+        :type: int
+        """
+        if not data_as_str:
+            return None
+
+        # Determine type for every cell
+        header_check_list = []
+        for row in data_as_str.split('\n'):
+            row_list = row.split(delimiter)
+            if len(row_list) == 0:
+                continue
+            header_check_list.append([])
+            for i in range(len(row_list)):
+                cell = row_list[i].strip()
+                cell_type = 'str'
+                try:
+                    float_cell = float(cell)
+                    if int(float_cell) == float_cell:
+                        cell_type = 'int'
+                    else:
+                        cell_type = 'float'
+                except ValueError:
+                    pass
+
+                if cell_type not in ['float', 'int']:
+                    if len(cell) == 0:
+                        cell_type = 'none'
+                    elif cell.isupper():
+                        cell_type = 'upstr'
+                header_check_list[-1].append(cell_type)
+
+        # Flags differences in types between each row (true/false)
+        potential_header = header_check_list[0]
+        differences = []
+        skip_rows = []
+        for i in range(1, len(header_check_list)):
+            differences.append([])
+            # check to see if row should be skiped due to error
+            if len(header_check_list[i]) > len(potential_header):
+                skip_rows.append(i - 1)
+                differences[i - 1] = [False] * len(header_check_list[i])
+            else:
+                for j in range(len(header_check_list[i])):
+                    diff_flag = False
+                    if header_check_list[i][j] != potential_header[j]:
+                        diff_flag = True
+                    differences[i - 1].append(diff_flag)
+            potential_header = header_check_list[i]
+
+        # Predicts the last row that could be the header, given the criteria
+        prior_len = 0
+        row_classic_header_ends = None
+        for i in range(0, len(differences)):
+            # Determine ratio of none in row, must be BELOW threshold
+            none = float(header_check_list[i].count("none")) / float(len(header_check_list[i]))
+            # Determine percent of differences between prior row, must be BELOW threshold
+            diff = float(differences[i].count(True)) / float(len(differences[i]))
+            # Determine percent of string, uppercase string or none in row, must be ABOVE threshold
+            rstr = float((header_check_list[i].count("str")
+                          + header_check_list[i].count("upstr")
+                          + header_check_list[i].count("none"))) / float(len(header_check_list[i]))
+            # Determines if the elements in the row is increasing or decreasing
```

Comment:
Do you mean to say "Determines if the **number of** elements in the row is increasing or decreasing?

Reply:
yeah, number of not-none elements, to be more specific. I'll change the comment when all code are approved

Reply:
This comment is still incorrect